connect/runtime/src/test/java/org/apache/kafka/connect/integration/RebalanceSourceConnectorsIntegrationTest.java

```diff
@@ -0,0 +1,374 @@
+        props.put("throughput", String.valueOf(1));
+        props.put("messages.per.poll", String.valueOf(10));
+        props.put(TOPICS_CONFIG, TOPIC_NAME);
+        props.put(KEY_CONVERTER_CLASS_CONFIG, StringConverter.class.getName());
+        props.put(VALUE_CONVERTER_CLASS_CONFIG, StringConverter.class.getName());
+
+        waitForCondition(() -> this.assertWorkersUp(3),
+                WORKER_SETUP_DURATION_MS, "Connect workers did not start in time.");
+
+        // start a source connector
+        IntStream.range(0, 4).forEachOrdered(
+            i -> {
+                try {
+                    connect.configureConnector(CONNECTOR_NAME + i, props);
+                } catch (IOException e) {
+                    throw new ConnectException(e);
+                }
+            });
+
+        waitForCondition(() -> this.assertConnectorAndTasksRunning(CONNECTOR_NAME + 3, 4).orElse(true),
+                CONNECTOR_SETUP_DURATION_MS, "Connector tasks did not start in time.");
+
+        // delete connector
+        connect.deleteConnector(CONNECTOR_NAME + 3);
+
+        waitForCondition(() -> !this.assertConnectorAndTasksRunning(CONNECTOR_NAME + 3, 4).orElse(true),
+                CONNECTOR_SETUP_DURATION_MS, "Connector tasks did not stop in time.");
+
+        waitForCondition(this::assertConnectorAndTasksAreBalanced,
+                WORKER_SETUP_DURATION_MS, "Connect and tasks are imbalanced between the workers.");
+    }
+
+    /**
+     * TODO: javadoc
+     */
+    @Test
+    public void testAddingWorker() throws Exception {
+        // create test topic
+        connect.kafka().createTopic(TOPIC_NAME, NUM_TOPIC_PARTITIONS);
+
+        // setup up props for the source connector
+        Map<String, String> props = new HashMap<>();
+        props.put(CONNECTOR_CLASS_CONFIG, MonitorableSourceConnector.class.getSimpleName());
+        props.put(TASKS_MAX_CONFIG, String.valueOf(NUM_TASKS));
+        props.put("throughput", String.valueOf(1));
+        props.put("messages.per.poll", String.valueOf(10));
+        props.put(TOPICS_CONFIG, TOPIC_NAME);
+        props.put(KEY_CONVERTER_CLASS_CONFIG, StringConverter.class.getName());
+        props.put(VALUE_CONVERTER_CLASS_CONFIG, StringConverter.class.getName());
+
+        waitForCondition(() -> this.assertWorkersUp(3),
+                WORKER_SETUP_DURATION_MS, "Connect workers did not start in time.");
+
+        // start a source connector
+        IntStream.range(0, 4).forEachOrdered(
+            i -> {
+                try {
+                    connect.configureConnector(CONNECTOR_NAME + i, props);
+                } catch (IOException e) {
+                    throw new ConnectException(e);
+                }
+            });
+
+        waitForCondition(() -> this.assertConnectorAndTasksRunning(CONNECTOR_NAME + 3, 4).orElse(false),
+                CONNECTOR_SETUP_DURATION_MS, "Connector tasks did not start in time.");
```

Comment:
We're starting 4 connectors, so should we check that all 4 are running?

Reply:
Here just confirming that they've started to go to the next phase (which is to remove a worker). Not an assertion of the connectors per se. But I agree we could tighten it in a subsequent iteration.